File: q3-bridge/bridge/extraction/q3-sdk.js
#!/usr/bin/env node
/**
 * Q3 Protocol - JavaScript SDK
 * S3-compatible client for Quantum Cube Storage
 * 
 * Usage:
 *   const Q3Client = require('./q3-sdk');
 *   const client = new Q3Client({ endpoint: 'http://localhost:7472' });
 *   await client.createBucket('my-bucket');
 *   await client.putObject('my-bucket', 'file.txt', 'Hello Q3!');
 */

class Q3Client {
    constructor(options = {}) {
        this.endpoint = options.endpoint || 'http://localhost:7472';
        this.apiBase = `${this.endpoint}/api/q3`;
    }

    /**
     * Create a new bucket
     */
    async createBucket(name, options = {}) {
        const response = await fetch(`${this.apiBase}/buckets`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ name, ...options })
        });

        const data = await response.json();
        if (!data.success) {
            throw new Error(data.error || 'Failed to create bucket');
        }

        return data.bucket;
    }

    /**
     * List all buckets
     */
    async listBuckets() {
        const response = await fetch(`${this.apiBase}/buckets`);
        const data = await response.json();
        return data.buckets || [];
    }

    /**
     * Upload an object
     */
    async putObject(bucket, key, data, metadata = {}) {
        const response = await fetch(`${this.apiBase}/${bucket}/${key}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                'X-Q3-Version': '1.0'
            },
            body: JSON.stringify({ data, metadata })
        });

        return await response.json();
    }

    /**
     * Download an object
     */
    async getObject(bucket, key) {
        const response = await fetch(`${this.apiBase}/${bucket}/${key}`);

        if (!response.ok) {
            throw new Error(`Object not found: ${bucket}/${key}`);
        }

        return await response.text();
    }

    /**
     * Delete an object
     */
    async deleteObject(bucket, key) {
        const response = await fetch(`${this.apiBase}/${bucket}/${key}`, {
            method: 'DELETE'
        });

        return await response.json();
    }

    /**
     * List objects in a bucket
     */
    async listObjects(bucket, prefix = '') {
        const url = new URL(`${this.apiBase}/${bucket}/objects`);
        if (prefix) url.searchParams.set('prefix', prefix);

        const response = await fetch(url);
        const data = await response.json();
        return data.objects || [];
    }

    /**
     * Get storage statistics
     */
    async getStats() {
        const response = await fetch(`${this.apiBase}/stats`);
        return await response.json();
    }
}

// CLI interface
if (require.main === module) {
    const args = process.argv.slice(2);
    const command = args[0];

    const client = new Q3Client();

    (async () => {
        try {
            switch (command) {
                case 'create-bucket':
                    const bucket = await client.createBucket(args[1]);
                    console.log('✅ Bucket created:', bucket.name);
                    break;

                case 'list-buckets':
                    const buckets = await client.listBuckets();
                    console.log('Buckets:', buckets.map(b => b.name).join(', '));
                    break;

                case 'put':
                    await client.putObject(args[1], args[2], args[3]);
                    console.log(`✅ Uploaded: ${args[1]}/${args[2]}`);
                    break;

                case 'get':
                    const data = await client.getObject(args[1], args[2]);
                    console.log(data);
                    break;

                case 'stats':
                    const stats = await client.getStats();
                    console.log('Q3 Storage Stats:', stats);
                    break;

                default:
                    console.log('Q3 Protocol - JavaScript SDK');
                    console.log('\nUsage:');
                    console.log('  node q3-sdk.js create-bucket <name>');
                    console.log('  node q3-sdk.js list-buckets');
                    console.log('  node q3-sdk.js put <bucket> <key> <data>');
                    console.log('  node q3-sdk.js get <bucket> <key>');
                    console.log('  node q3-sdk.js stats');
            }
        } catch (error) {
            console.error('Error:', error.message);
            process.exit(1);
        }
    })();
}

module.exports = Q3Client;

File: q3-bridge/bridge/extraction/q3_sdk.py
#!/usr/bin/env python3
"""
Q3 Protocol - Python SDK
S3-compatible client for Quantum Cube Storage

Usage:
    from q3_sdk import Q3Client
    
    client = Q3Client(endpoint='http://localhost:7472')
    client.create_bucket('my-bucket')
    client.put_object('my-bucket', 'file.txt', b'Hello Q3!')
    data = client.get_object('my-bucket', 'file.txt')
"""

import requests
import json
from typing import Dict, List, Optional, Union

class Q3Client:
    """Q3 Protocol Python Client"""
    
    def __init__(self, endpoint: str = 'http://localhost:7472'):
        self.endpoint = endpoint
        self.api_base = f'{endpoint}/api/q3'
        self.session = requests.Session()
        self.session.headers.update({
            'X-Q3-Version': '1.0'
        })
    
    def create_bucket(self, name: str, **options) -> Dict:
        """Create a new Q3 bucket"""
        response = self.session.post(
            f'{self.api_base}/buckets',
            json={'name': name, **options}
        )
        response.raise_for_status()
        data = response.json()
        
        if not data.get('success'):
            raise Exception(data.get('error', 'Failed to create bucket'))
        
        return data['bucket']
    
    def list_buckets(self) -> List[Dict]:
        """List all buckets"""
        response = self.session.get(f'{self.api_base}/buckets')
        response.raise_for_status()
        return response.json().get('buckets', [])
    
    def put_object(self, bucket: str, key: str, data: Union[str, bytes], 
                   metadata: Optional[Dict] = None) -> Dict:
        """Upload an object to a bucket"""
        if isinstance(data, bytes):
            data = data.decode('utf-8')
        
        response = self.session.put(
            f'{self.api_base}/{bucket}/{key}',
            json={'data': data, 'metadata': metadata or {}}
        )
        response.raise_for_status()
        return response.json()
    
    def get_object(self, bucket: str, key: str) -> bytes:
        """Download an object from a bucket"""
        response = self.session.get(f'{self.api_base}/{bucket}/{key}')
        response.raise_for_status()
        return response.content
    
    def delete_object(self, bucket: str, key: str) -> Dict:
        """Delete an object from a bucket"""
        response = self.session.delete(f'{self.api_base}/{bucket}/{key}')
        response.raise_for_status()
        return response.json()
    
    def list_objects(self, bucket: str, prefix: str = '') -> List[Dict]:
        """List objects in a bucket"""
        params = {'prefix': prefix} if prefix else {}
        response = self.session.get(
            f'{self.api_base}/{bucket}/objects',
            params=params
        )
        response.raise_for_status()
        return response.json().get('objects', [])
    
    def get_stats(self) -> Dict:
        """Get storage statistics"""
        response = self.session.get(f'{self.api_base}/stats')
        response.raise_for_status()
        return response.json()


# CLI interface
if __name__ == '__main__':
    import sys
    
    if len(sys.argv) < 2:
        print('Q3 Protocol - Python SDK')
        print('\nUsage:')
        print('  python3 q3_sdk.py create-bucket <name>')
        print('  python3 q3_sdk.py list-buckets')
        print('  python3 q3_sdk.py put <bucket> <key> <data>')
        print('  python3 q3_sdk.py get <bucket> <key>')
        print('  python3 q3_sdk.py stats')
        sys.exit(1)
    
    command = sys.argv[1]
    client = Q3Client()
    
    try:
        if command == 'create-bucket':
            bucket = client.create_bucket(sys.argv[2])
            print(f'✅ Bucket created: {bucket["name"]}')
        
        elif command == 'list-buckets':
            buckets = client.list_buckets()
            print('Buckets:', ', '.join(b['name'] for b in buckets))
        
        elif command == 'put':
            result = client.put_object(sys.argv[2], sys.argv[3], sys.argv[4])
            print(f'✅ Uploaded: {sys.argv[2]}/{sys.argv[3]}')
        
        elif command == 'get':
            data = client.get_object(sys.argv[2], sys.argv[3])
            print(data.decode('utf-8'))
        
        elif command == 'stats':
            stats = client.get_stats()
            print('Q3 Storage Stats:', json.dumps(stats, indent=2))
        
        else:
            print(f'Unknown command: {command}')
            sys.exit(1)
    
    except Exception as e:
        print(f'Error: {e}', file=sys.stderr)
        sys.exit(1)
